proxylib/memcached/text/parser.go

```diff
@@ -0,0 +1,331 @@
+		// in watch mode we pass all replied lines
+		return proxylib.PASS, linefeed + 2
+	}
+
+	switch {
+	case p.isErrorReply(tokens[0]) ||
+		p.isCommandStorage(intent.command) ||
+		p.isCommandDelete(intent.command) ||
+		p.isCommandIncrDecr(intent.command) ||
+		bytes.Equal(intent.command, []byte("touch")) ||
+		bytes.Equal(intent.command, []byte("slabs")) ||
+		bytes.Equal(intent.command, []byte("lru")) ||
+		bytes.Equal(intent.command, []byte("flush_all")) ||
+		bytes.Equal(intent.command, []byte("cache_memlimit")) ||
+		bytes.Equal(intent.command, []byte("version")) ||
+		bytes.Equal(intent.command, []byte("misbehave")):
+
+		// passing one line of reply
+		p.connection.Log(cilium.EntryType_Response, logEntry)
+		return proxylib.PASS, linefeed + 2
+	case p.isCommandRetrieval(intent.command) ||
+		bytes.Equal(intent.command, []byte("stats")):
+		t, nBytes := p.untilEnd(data)
+		if t == proxylib.PASS {
+			p.connection.Log(cilium.EntryType_Response, logEntry)
+			p.replyQueue = p.replyQueue[1:]
+		}
+		return t, nBytes
+	case bytes.Equal(intent.command, []byte("lru_crawler")):
+		// check if it's response line
+		if bytes.Equal(tokens[0], []byte("OK")) ||
+			bytes.Equal(tokens[0], []byte("BUSY")) ||
+			bytes.Equal(tokens[0], []byte("BADCLASS")) {
+			p.connection.Log(cilium.EntryType_Response, logEntry)
+			return proxylib.PASS, linefeed + 2
+		}
+
+		t, nBytes := p.untilEnd(data)
+		if t == proxylib.PASS {
+			p.connection.Log(cilium.EntryType_Response, logEntry)
+			p.replyQueue = p.replyQueue[1:]
+		}
+		return t, nBytes
+	}
+	log.Error("Could not parse text memcache frame")
+	return proxylib.ERROR, 0
+}
+
+func (p *Parser) untilEnd(data []byte) (proxylib.OpType, int) {
+	// TODO: optimise this to not ask per byte, but take VALUES lines into account
+	endIndex := bytes.Index(data, []byte("END\r\n"))
+	if endIndex > 0 {
+		return proxylib.PASS, endIndex + 5
+	}
+	return proxylib.MORE, 1
+}
+
+func (p *Parser) isCommandRetrieval(cmd []byte) bool {
+	return bytes.HasPrefix(cmd, []byte("get")) ||
+		bytes.HasPrefix(cmd, []byte("gat"))
+}
+
+func (p *Parser) isCommandStorage(cmd []byte) bool {
+	return bytes.Equal(cmd, []byte("set")) ||
+		bytes.Equal(cmd, []byte("add")) ||
+		bytes.Equal(cmd, []byte("replace")) ||
+		bytes.Equal(cmd, []byte("append")) ||
+		bytes.Equal(cmd, []byte("prepend")) ||
+		bytes.Equal(cmd, []byte("cas"))
+}
+
+func (p *Parser) isCommandDelete(cmd []byte) bool {
+	return bytes.Equal(cmd, []byte("delete"))
+}
+
+func (p *Parser) isCommandIncrDecr(cmd []byte) bool {
+	return bytes.Equal(cmd, []byte("incr")) ||
+		bytes.Equal(cmd, []byte("decr"))
+}
+
+func (p *Parser) isErrorReply(firstToken []byte) bool {
+	return bytes.Equal(firstToken, []byte("ERROR")) ||
+		bytes.Equal(firstToken, []byte("CLIENT_ERROR")) ||
+		bytes.Equal(firstToken, []byte("SERVER_ERROR"))
+}
+
+func (p *Parser) injectFromQueue() bool {
+	if len(p.replyQueue) > 0 {
+		if p.replyQueue[0].requestID == p.replyCount+1 && p.replyQueue[0].denied {
```

Comment:
If we push an element to the queue for each expected reply in addition to the denied replies, then the comparison of the requestID is no longer needed? I.e., if the head of the queue is `denied` then we have already seen all the replies that caused the denied reply be queued in the first place and can just inject the denied message?

Also, we should loop here if there happens to be multiple denied messages at the head of the queue?

Reply:
done